python/lsst/ts/salobj/validator.py

```diff
@@ -36,6 +36,30 @@ class DefaultingValidator:
 
     Notes
     -----
+    Default values are handled at most 2 levels deep in an object hierarchy.
+    For deeper hierarchies, set the default at a higher level.
+    For example::
+
+        type: object
+        properties:
+          number1:
+            type: number
+            default: 1
+          subdict1:
+            type: object
+            properties:
+              number2:
+                type: number
+                default: 2
+              subdict2:
+                type: object
+                properties:
+                  number3:
+                    type: number
+                    # default is ignored this deep; set it at a higher level
+                default:
+                  number3: 3
+
```

Comment:
Thanks for adding this. That's very useful.

The issue I was struggling with is that skipping a default on one level, makes that level and all sub-levels completely get removed when the configuration is loaded. Here is what I tried. Note that the `devices` property does not have a default value.

properties:
  devices:
    type: array
    default:
    items:
      type: object
      properties:
        name:
          description: Name of the sensor
          type: string
          default: EssTemperature4Ch
        channels:
          description: Number of channels
          type: integer
          default: 4
        type:
          description: Type of the device
          type: string
          enum:
          - FTDI
          - Serial
          default: FTDI

Sorry for the formatting. GitHub doesn't allow for extensive formatting in comments. Maybe we should discuss this on Slack?

Reply:
I commented on the Jira ticket, but I think that question may be getting a bit too detailed.